src/shared/install.c

```diff
@@ -1720,19 +1720,65 @@ static int install_info_symlink_alias(
         assert(config_path);
 
         STRV_FOREACH(s, i->aliases) {
+
                 _cleanup_free_ char *alias_path = NULL, *dst = NULL;
 
                 q = install_full_printf(i, *s, &dst);
                 if (q < 0)
                         return q;
 
+                if (unit_name_is_valid(i->name, UNIT_NAME_TEMPLATE) && !unit_name_is_valid(dst, UNIT_NAME_TEMPLATE) && !i->default_instance)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Alias for templates without DefaultInstance must be templates.");
+
+                if(unit_name_is_valid(i->name, UNIT_NAME_INSTANCE) || unit_name_is_valid(i->name, UNIT_NAME_TEMPLATE)) {
+                        bool is_valid = false;
+                        if (unit_name_is_valid(dst, UNIT_NAME_TEMPLATE))
+                                is_valid = true;
+                        if (unit_name_is_valid(dst, UNIT_NAME_INSTANCE) && strstr(*s,"%i"))
+                                is_valid = true;
+                        if (unit_name_is_valid(dst, UNIT_NAME_INSTANCE) && strstr(*s,"%I"))
+                                is_valid = true;
+                        if (unit_name_is_valid(dst, UNIT_NAME_INSTANCE) && strstr(*s,"%f"))
+                                is_valid = true;
```

Comment:
I don't think this part is useful. First of all, we don't want to build a list of possible specifiers here, and it also matches %%i, so it doesn't work properly anyway. And we don't really care how the destination instance is built, only that the result is valid.

Maybe something like this:
```C
  _cleanup_free_ char *src_inst = NULL;
  int src_type = unit_name_to_instance(src, &src_inst);
  ...

  _cleanup_free_ char *dst_inst = NULL;
  int dst_type = unit_name_to_instance(dst, &dst_inst);

        ...
  else if (src_type == UNIT_NAME_INSTANCE)) {
       if (dst_type == UNIT_NAME_PLAIN)
                return log_error_errno(...);
       if (dst_type == UNIT_NAME_INSTANCE &&
           !streq_ptr(dst_inst ?: i->default_instance, dst_instance))
                return log_error_errno(...);
 };
```

Reply:
reworked to something similar to what you offer, but I still need to use strstr not streq since the point is to catch `Alias=target@my-%i` so I check if the resulting instance contains the original instance, not "is equal to"

Reply:
No. The instance value must be unique. It is not OK to alias to something with a different instance value.

Reply:
I'm not sure I understand... you want to allow `template@%i` but not `template@my-%i` ?  in that case, why allow any expansion at all ?

The point was to not break wpa_supplicant's services, or so I thought...
i'm a bit confused on what you want to allow/disallow here... can you clarify ?

Reply:
Oh, I see that https://github.com/systemd/systemd/pull/13990#issuecomment-561301805 now includes a row with differing instance attributes. I don't know how I didn't notice it earlier. All those cases where the instance is different between two aliases must be rejected.

Reply:
sorry but I still don't understand what you mean. can you give me a case where my patch (or just the table in the comment) is not what you want and tell me what you want instead ?

I might be a little out of touch on that one, but I want to be pretty sure before I rework the patch.